vid/src/avid_m/proofs.rs

```diff
@@ -127,6 +133,107 @@ impl MalEncodingProof {
     }
 }
 
+/// A proof of incorrect encoding for a namespace.
+/// It consists of the index of the namespace, the merkle proof of the namespace payload against the namespaced VID commitment,
+/// and the proof of incorrect encoding.
+#[derive(Clone, Debug, Serialize, Deserialize, Eq, PartialEq)]
+pub struct NsAvidMBadEncodingProof {
+    /// The index of the namespace.
+    pub ns_index: usize,
+    /// The commitment of the namespaced VID.
+    pub ns_commit: AvidMCommit,
+    /// The outer merkle proof of the namespace against the namespaced VID commitment.
+    pub ns_mt_proof: MerkleProof,
+    /// The proof of incorrect encoding.
+    pub ns_proof: AvidMBadEncodingProof,
+}
+
+impl NsAvidMScheme {
+    /// Generate a proof of incorrect encoding for a namespace.
+    pub fn proof_of_incorrect_encoding_for_namespace(
+        param: &AvidMParam,
+        ns_index: usize,
+        commit: &NsAvidMCommit,
+        shares: &[NsAvidMShare],
+    ) -> VidResult<NsAvidMBadEncodingProof> {
+        if shares.is_empty() {
+            return Err(VidError::InsufficientShares);
+        }
+        if shares.iter().any(|share| !share.contains_ns(ns_index)) {
+            return Err(VidError::IndexOutOfBound);
+        }
+        let mt = MerkleTree::from_elems(
+            None,
+            shares[0].ns_commits().iter().map(|commit| commit.commit),
+        )?;
+        if mt.commitment() != commit.commit {
+            return Err(VidError::InvalidShare);
+        }
+        let (ns_commit, ns_mt_proof) = mt
+            .lookup(ns_index as u64)
+            .expect_ok()
+            .expect("MT lookup shouldn't fail");
+        let ns_commit = AvidMCommit { commit: *ns_commit };
+        let shares = shares
+            .iter()
+            .filter_map(|share| share.inner_ns_share(ns_index))
+            .collect::<Vec<_>>();
+        Ok(NsAvidMBadEncodingProof {
+            ns_index,
+            ns_commit,
+            ns_mt_proof,
+            ns_proof: AvidMScheme::proof_of_incorrect_encoding(param, &ns_commit, &shares)?,
+        })
+    }
+
+    /// Generate a proof of incorrect encoding.
+    pub fn proof_of_incorrect_encoding(
+        param: &AvidMParam,
+        commit: &NsAvidMCommit,
+        shares: &[NsAvidMShare],
+    ) -> VidResult<NsAvidMBadEncodingProof> {
+        if shares.is_empty() {
+            return Err(VidError::InsufficientShares);
+        }
+        for ns_index in 0..shares[0].ns_commits().len() {
+            let result =
+                Self::proof_of_incorrect_encoding_for_namespace(param, ns_index, commit, shares);
```

Comment:
when will someone use this method? it feels strange to shares of a specific namespace, but not specifying the namespace index, and ask this method to enumerate.

Reply:
There's a problem about current implementation. If any namespace within this block is malformed, you cannot generate a good namespace proof for any namespace. So if you failed generating the namespace proof you don't know which namespace is the cause. You are forced to use this method tho.